modules/ocf/files/ssl/ocf-lets-encrypt

```diff
@@ -0,0 +1,253 @@
+    output = subprocess.check_output((
+        'openssl', 'x509', '-enddate', '-noout',
+        '-in', str(cert_path),
+    ))
+    m = re.match(b'notAfter=(.+)$', output)
+    assert m, output
+    return dateutil.parser.parse(m.group(1).decode('ascii'))
+
+
+@contextmanager
+def make_csr(domain, key):
+    _, csr_path = tempfile.mkstemp()
+    try:
+        subprocess.check_call((
+            'openssl', 'req', '-new', '-sha256',
+            '-key', key,
+            '-subj', '/CN={}'.format(domain),
+            '-out', csr_path,
+        ))
+        yield csr_path
+    finally:
+        os.remove(csr_path)
+
+
+def acme_tiny(domain, key):
+    with make_csr(domain, key) as csr_path:
+        # TODO: use subprocess.run when we get python3.5
+        proc = subprocess.Popen(
+            (
+                'acme-tiny',
+                '--account-key', str(LE_ACCOUNT_KEY),
+                '--csr', csr_path,
+                '--acme-dir', str(ACME_BASE_DIR / '.well-known' / 'acme-challenge'),
+                # '--ca', 'https://acme-staging.api.letsencrypt.org',
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = proc.communicate()
+    return stdout, stderr, proc.returncode
+
+
+def write_cert(cert, cert_path):
+    """Write certificate file for a domain."""
+    _, temp_cert_path = tempfile.mkstemp(dir=str(cert_path.parent), prefix='.tmp')
+    try:
+        with open(temp_cert_path, 'wb') as f:
+            f.write(cert)
+        # atomic overwrite in case of errors
+        os.chmod(temp_cert_path, 0o644)
+        shutil.move(temp_cert_path, str(cert_path))
+    finally:
+        try:
+            os.remove(temp_cert_path)
+        except FileNotFoundError:
+            pass
+
+
+def acquire(request, key, dry_run=False):
+    """Makes a Let's Encrypt certificate request."""
+    if not dry_run:
+        try:
+            cert, stderr, returncode = acme_tiny(request.domain, key)
+            assert returncode == 0, returncode
+            write_cert(cert, request.cert_path)
+            if request.has_cert:
+                print('cert renewed for: ' + request.domain)
+            else:
+                print('cert acquired for: ' + request.domain)
+            return True
+        except:
+            print('An error occured!', file=sys.stderr)
+            print('stderr from acme-tiny:', file=sys.stderr)
+            for line in stderr.decode().split('\n'):
+                print('>', line, file=sys.stderr)
+            raise
+    else:
+        return False
+
+
+@contextmanager
+def maybe_start_webserver(dry_run):
+    if not dry_run and socket.socket().connect_ex(('127.0.0.1', 80)) != 0:
+        proc = subprocess.Popen(
+            ('python3', '-m', 'http.server', '80'),
+            cwd=str(ACME_BASE_DIR),
+        )
+
+        # Wait for server to start up before proceeding
+        while socket.socket().connect_ex(('127.0.0.1', 80)) != 0:
```

Comment:
Let's hope the server always manages to start up, or else this script will hang forever...

Reply:
Oh, thanks, it should error if the process dies.